### ISSUE TITLE
not work for OrtCUDAProviderOptionsV2 to release GPU memory in C++

### ISSUE THREAD

**Issue by veelion** [microsoft/onnxruntime#12748]
Thanks @Z-XQ for the explanation.

The GPU memory is backed by a memory pool (arena) and we have a config knob to shrink the arena (de-allocated unused memory chunks). 

Not sure if we have enough tools to accomplish this in Python just yet. The best way to use this feature in C++ is to:

1) Not allocate weights memory through the arena: See [here](https://github.com/microsoft/onnxruntime/blob/651955d3c91be9a1cbb3ce29360758c8347016ad/onnxruntime/test/shared_lib/test_inference.cc#L1731)

2) Configure the arena to have high enough initial chunk to support most Run() calls. See "initial_chunk_size_bytes" [here](https://github.com/microsoft/onnxruntime/blob/651955d3c91be9a1cbb3ce29360758c8347016ad/onnxruntime/test/shared_lib/test_inference.cc#L1758)

3) Finally, configure the arena to shrink on every Run(). See [here](https://github.com/microsoft/onnxruntime/blob/651955d3c91be9a1cbb3ce29360758c8347016ad/onnxruntime/test/shared_lib/test_inference.cc#L1777). This will keep the initial chunk allocated but de-allocate any unused chunk remaining after the Run() call ends.

For example, if the initial chunk size is set as 500MB, the first Run() will allocate 500MB + any additional chunks required to service the Run() call. The additional chunks will get de-allocated after Run() and only keep 500MB of memory allocated.  It is important to not allocate weights (initializers) memory through the arena as that complicates the shrinkage. Hence, step (1).

_Originally posted by @hariharans29 in https://github.com/microsoft/onnxruntime/issues/9509#issuecomment-951546580_

**Comment by veelion** [comment]
This method used the legacy OrtCUDAProviderOptions, but failed to OrtCUDAProviderOptionsV2. Because V2 does not let us set value to its attributes, which means the flowing code will fail:
`cuda_option_v2.default_memory_arena_cfg = arena_cfg`

**Comment by hariharans29** [comment]
That's right. This is a known gap in the C API. The `OrtCUDAProviderOptionsV2` only supports string key/value configurations. There needs to be a new API to append an `OrtArenaCfg` to the opaque struct `OrtCUDAProviderOptionsV2`.

Currently, the only way to use `OrtArenaCfg` to configure the allocator of the CUDA EP is to use the legacy `OrtCUDAProviderOptions`.

CC: @pranavsharma @petermcaughan

**Comment by veelion** [comment]
Thanks for your reply, looking forward to the new API.

**Comment by pskiran1** [comment]
@hariharans29, @pranavsharma @petermcaughan, any update on this? do you have a plan to support `default_memory_arena_cfg` 
 in `OrtCUDAProviderOptionsV2`?
When I attempted the following, I encountered no errors, and the ArenaCfg did not work.

```
std::vector<const char*> keys;
std::vector<size_t> values;

keys.push_back("initial_chunk_size_bytes");
values.push_back(0);
keys.push_back("initial_chunk_size_bytes");
values.push_back(67108864);

OrtArenaCfg* arena_cfg = nullptr;
ort_api->CreateArenaCfgV2(keys.data(), values.data(), keys.size(), &arena_cfg);
std::unique_ptr<OrtArenaCfg, decltype(ort_api->ReleaseArenaCfg)>rel_arena_cfg(arena_cfg, ort_api->ReleaseArenaCfg);
ort_api->UpdateCUDAProviderOptionsWithValue(rel_cuda_options.get(), "default_memory_arena_cfg", rel_arena_cfg.get());
```

Output logs:
```
[I:onnxruntime:log, bfc_arena.cc:29 BFCArena] Creating BFCArena for Cuda with following configs: initial_chunk_size_bytes: 1048576 max_dead_bytes_per_chunk: 134217728 initial_growth_chunk_size_bytes: 2097152 max_power_of_two_extend_bytes: 1073741824 memory limit: 18446744073709551615 arena_extend_strategy: 0

[I:onnxruntime:log, bfc_arena.cc:29 BFCArena] Creating BFCArena for CudaPinned with following configs: initial_chunk_size_bytes: 1048576 max_dead_bytes_per_chunk: 134217728 initial_growth_chunk_size_bytes: 2097152 max_power_of_two_extend_bytes: 1073741824 memory limit: 18446744073709551615 arena_extend_strategy: 0

[I:onnxruntime:log, bfc_arena.cc:29 BFCArena] Creating BFCArena for Cpu with following configs: initial_chunk_size_bytes: 1048576 max_dead_bytes_per_chunk: 134217728 initial_growth_chunk_size_bytes: 2097152 max_power_of_two_extend_bytes: 1073741824 memory limit: 18446744073709551615 arena_extend_strategy: 0
```
